### PR TITLE
900 add filter search

### LINKED ISSUE #900: Add filter search to searchable collections
In the C# library, you can do a filter search like:

```
var accountsContainingLuc = await myApplication
    .GetAccounts()
    .Filter("Luc")
    .ToListAsync();
```

The closest we currently have in the JSDK is:

```
Map<String, Object> queryParams = new HashMap<>();
queryParams.put("q", "Luc");
AccountList accounts = application.getAccounts(queryParams);
```

We should add a method to our collection search objects (Like `Accounts`) so that we could do a filter search like:

```
AccountList accounts = application.getAccounts(
   Accounts.filter("Luc")
);
```

Behind the scenes, it would do the `q=Luc` request.

### PR DESCRIPTION
**Pull request by mzumbadoate**
fixes #900

### CODE REVIEW

**Comment by mrioan** [comment]
As described [here](https://docs.stormpath.com/rest/product-guide/latest/reference.html?highlight=filter#search) the filter-based search supports sorting, pagination, and link expansion. Please update the new IT that you added in order to check that at least sorting and expansion works fine when `filter()` is used. Something like:
```
application.getAccounts(Accounts.where(Accounts.filter(myFilterText)).orderBySurname().descending().withGroups(2);
```
Additionally, please document the `filter()` method. Something like this:
```
This operation will trigger the backend to carry out a case-insensitive matching query on all viewable attributes in all the resources in the Collection. 
So the following query:

application.getAccounts(Accounts.where(Accounts.filter("Joe"));

Returns all Accounts where:

* Each Account exists inside an AccountStore belonging to the Application `application`
* The Account’s givenName equals or contains “joe” (case insensitive) OR
* The Account’s middlename equals or contains “joe” (case insensitive) OR
* The Account’s email equals or contains “joe” (case insensitive) OR
And so on
```

**Comment by mrioan** [comment]
@mzumbadoate back to you

**Comment by mzumbadoate** [comment]
@dogeared since you created this ticket (#900) can you confirm @mrioan comment please

**Comment by lhazlewood** [comment]
FWIW, this PR should enable this for *all* collections that support filter queries (which is practically all of them) - not just application accounts or groups.

**Comment by lhazlewood** [comment]
Also, use of a filter is another `Criterion` - it is a condition/part of a query, so it probably should be implemented as an instance of the `Criterion` interface, not a map.

**Comment by lhazlewood** [comment]
I added this back to ready since there are a lot more changes expected and @mzumbadoate is out on vacation.

**Comment by mzumbadoate** [comment]
@mrioan @lhazlewood This is ready for a new round of reviews

File: api/src/main/java/com/stormpath/sdk/directory/Directories.java [hunk 1]
```diff
@@ -222,6 +222,28 @@ public static EqualsExpressionFactory status() {
...
+     * Returns all Directories where:
+     *
+     * Each Directory exists inside an DirectoryStore belonging to the Tenant `tenant`
+     * The Directories’s name equals or contains “FilterDirectories” (case insensitive) OR
+     * The Directories’s description equals or contains “FilterDirectories” (case insensitive) OR
+     * And so on
+     *
+     * @param value The value to search for
+     * @return A Criterion representing the filter query
+     * @since 1.1.0
```
Comment (jarias) [inline api/src/main/java/com/stormpath/sdk/directory/Directories.java]: Versions need to be updated to 1.2.0

File: api/src/main/java/com/stormpath/sdk/group/Groups.java [hunk 2]
```diff
@@ -190,6 +190,27 @@ public static CreateGroupRequestBuilder newCreateRequestFor(Group group) {
...
+     * Returns all Groups where:
+     *
+     * Each Group exists inside an GroupStore belonging to the Application `application`
+     * The Group’s name equals or contains “Group01” (case insensitive) OR
+     * The Group’s description equals or contains “Group01” (case insensitive) OR
+     * And so on
+     *
+     * @param value The value to search for
+     * @return A Criterion representing the filter query
+     * @since 1.1.0
```
Comment (jarias) [inline api/src/main/java/com/stormpath/sdk/group/Groups.java]: 1.2.0

File: api/src/main/java/com/stormpath/sdk/application/Applications.java [hunk 3]
```diff
@@ -297,4 +297,26 @@ public static OAuthApiRequestAuthenticator oauthRequestAuthenticator(Application
...
+     * Returns all Applications where:
+     *
+     * Each Application exists inside an ApplicationStore belonging to the Tenant `tenant`
+     * The Application’s name equals or contains “Java SDK IT App” (case insensitive) OR
+     * The Applications’s description equals or contains “Java SDK IT App” (case insensitive) OR
+     * And so on
+     *
+     * @param value The value to search for
+     * @return A Criterion representing the filter query
+     * @since 1.1.0
```
Comment (jarias) [inline api/src/main/java/com/stormpath/sdk/application/Applications.java]: 1.2.0

File: api/src/main/java/com/stormpath/sdk/account/Accounts.java [hunk 4]
```diff
@@ -296,6 +296,29 @@ public static DateExpressionFactory modifiedAt(){
...
+     *
+     * Each Account exists inside an AccountStore belonging to the Application `application`
+     * The Account’s givenName equals or contains “joe” (case insensitive) OR
+     * The Account’s middlename equals or contains “joe” (case insensitive) OR
+     * The Account’s email equals or contains “joe” (case insensitive) OR
+     * And so on
+     *
+     * @param value The value to search for
+     * @return A Criterion representing the filter query
+     * @since 1.1.0
```
Comment (jarias) [inline api/src/main/java/com/stormpath/sdk/account/Accounts.java]: 1.2.0

File: api/src/main/java/com/stormpath/sdk/organization/Organizations.java [hunk 5]
```diff
@@ -225,6 +225,28 @@ public static EqualsExpressionFactory status() {
...
+     * Returns all Organizations where:
+     *
+     * Each Organization exists inside an OrganizationStore belonging to the Tenant `tenant`
+     * The Organizations’s name equals or contains “FilterOrganizations” (case insensitive) OR
+     * The Organizations’s description equals or contains “FilterOrganizations” (case insensitive) OR
+     * And so on
+     *
+     * @param value The value to search for
+     * @return A Criterion representing the filter query
+     * @since 1.1.0
```
Comment (jarias) [inline api/src/main/java/com/stormpath/sdk/organization/Organizations.java]: 1.2.0

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/client/DirectoryIT.groovy [hunk 6]
```diff
@@ -108,6 +108,55 @@ class DirectoryIT extends ClientIT {
     }
 
     /**
+     * @since 1.1.0
```
Comment (jarias) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/client/DirectoryIT.groovy]: 1.2.0

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/account/AccountIT.groovy [hunk 7]
```diff
@@ -1060,4 +1060,71 @@ class AccountIT extends ClientIT {
         return secret
     }
 
+    /**
+     * @since 1.1.0
```
Comment (jarias) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/account/AccountIT.groovy]: 1.2.0

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy [hunk 8]
```diff
@@ -245,6 +245,60 @@ class ApplicationIT extends ClientIT {
         assertFalse list.iterator().hasNext() //no results
     }
 
+    /**
+     * @since 1.1.0
```
Comment (jarias) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy]: 1.2.0

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/group/GroupIT.groovy [hunk 9]
```diff
@@ -179,4 +180,54 @@ class GroupIT extends ClientIT {
             assertEquals e.getMessage(), "The specified account does not belong to this Group."
         }
     }
+
+    /**
+     * @since 1.1.0
```
Comment (jarias) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/group/GroupIT.groovy]: 1.2.0

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/organization/OrganizationIT.groovy [hunk 10]
```diff
@@ -101,6 +101,66 @@ class OrganizationIT extends ClientIT {
         deleteOnTeardown(retrievedOrg)
     }
 
+    /**
+     * @since 1.1.0
```
Comment (jarias) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/organization/OrganizationIT.groovy]: 1.2.0

**Comment by mzumbadoate** [comment]
@jarias version updated to 1.2.0 and also a merge with master is done
The CI is crashing because of the know issue https://github.com/stormpath/stormpath-sdk-java/issues/1027

**Comment by jarias** [comment]
LGTM @mzumbadoate just check the CI build, probably you need to merge master into your branch to fix the spec properties test.

**Pull request merged**